lib/rubygems/commands/rebuild_command.rb

```diff
@@ -0,0 +1,262 @@
+
+require "date"
+require "digest"
+require "fileutils"
+require_relative "../package"
+
+class Gem::Commands::RebuildCommand < Gem::Command
+  DATE_FORMAT = "%Y-%m-%d %H:%M:%S.%N Z"
+
+  def initialize
+    super "rebuild", "Attempt to reproduce a build of a gem."
+
+    add_option "--force", "Skip validation of the spec." do |_value, options|
+      options[:force] = true
+    end
+
+    add_option "--strict", "Consider warnings as errors when validating the spec." do |_value, options|
+      options[:strict] = true
+    end
+
+    add_option "--source GEM_SOURCE", "Specify the source to download the gem from." do |value, options|
+      options[:source] = value
+    end
+
+    add_option "--original GEM_FILE", "Specify a local file to compare against (instead of downloading it)." do |value, options|
+      options[:original_gem_file] = value
+    end
+
+    add_option "--gemspec GEMSPEC_FILE", "Specify the name of the gemspec file." do |value, options|
+      options[:gemspec_file] = value
+    end
+
+    add_option "-C PATH", "Run as if gem build was started in <PATH> instead of the current working directory." do |value, options|
+      options[:build_path] = value
+    end
+  end
+
+  def arguments # :nodoc:
+    "GEM_NAME      gem name on gem server\n" \
+    "GEM_VERSION   gem version you are attempting to rebuild"
+  end
+
+  def description # :nodoc:
+    <<-EOF
+The rebuild command allows you to (attempt to) reproduce a build of a gem
+from a ruby gemspec.
+
+This command assumes the gemspec can be built with the `gem build` command.
+If you use either `gem build` or `rake build`/`rake release` to build/release
+a gem, it is a potential candidate.
+
+If the gem includes top-level files that change frequently (e.g. Gemfile.lock),
+it may require more effort to reproduce a build. For example, it might require
+more precisely matched versions of Ruby, RubyGems, or Bundler to be used.
+
+An example of reproducing a gem build:
+
+  $ pwd
+  /usr/home/puppy/test/rebuild
+  $ ls -a
+  ./  ../
+  $ git clone --branch v12.0.2 https://github.com/duckinator/okay.git
+  $ gem rebuild -C ./okay --gemspec okay.gemspec okay 12.0.2
+  Fetching okay-12.0.2.gem
+  Downloaded okay version 12.0.2 as /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem.
+    Successfully built RubyGem
+    Name: okay
+    Version: 12.0.2
+    File: okay-12.0.2.gem
+
+  Built at: 2023-08-31 21:39:02 EDT (1693532342)
+  Original build saved to:   /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem
+  Reproduced build saved to: /usr/home/puppy/test/rebuild/rebuild/new/okay-12.0.2.gem
+  Working directory: ./okay
+
+  Hash comparison:
+    38a8bd78ce10bc19189ead0b56fa490d03788a2f926a6481b2f1f5d5fa5ab75b      /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem
+    38a8bd78ce10bc19189ead0b56fa490d03788a2f926a6481b2f1f5d5fa5ab75b      /usr/home/puppy/test/rebuild/rebuild/new/okay-12.0.2.gem
+
+  SUCCESS - original and rebuild hashes matched
+    EOF
```

Comment:
Do we want to include the full output of the current implementation in the command documentation/description? I think it's likely to get out of date and become confusing?

Reply:
It probably makes sense to pull this out and include it (with some other information) on <https://guides.rubygems.org>.

Reply:
This was mainly there so I had a practical example to reference. I think <https://smallest.dog/resources/reproducible-gem-builds/> has taken over that purpose, however, so I'm removing it from the help text.